components/cli/pkg/registry/registry.go

```diff
@@ -0,0 +1,181 @@
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http:www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package registry
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"io/ioutil"
+
+	"github.com/docker/distribution/manifest"
+	"github.com/docker/distribution/manifest/schema1"
+	"github.com/docker/libtrust"
+	registry2 "github.com/nokia/docker-registry-client/registry"
+	"github.com/opencontainers/go-digest"
+
+	"github.com/cellery-io/sdk/components/cli/pkg/image"
+	"github.com/cellery-io/sdk/components/cli/pkg/util"
+)
+
+type Registry interface {
+	Pull(parsedCellImage *image.CellImage, username string, password string) ([]byte, error)
+	Push(parsedCellImage *image.CellImage, fileBytes []byte, username, password string) error
+	Out() io.Writer
+}
+
+type CelleryRegistry struct {
+	hub *registry2.Registry
+}
+
+func NewCelleryRegistry() *CelleryRegistry {
+	registry := &CelleryRegistry{}
+	return registry
+}
+
+func (registry *CelleryRegistry) Push(parsedCellImage *image.CellImage, fileBytes []byte, username, password string) error {
+	fmt.Println("\nConnecting to ", util.Bold(parsedCellImage.Registry))
+	// Initiating a connection to Cellery Registry
+	hub, err := registry2.New("https://"+parsedCellImage.Registry, username, password)
+	if err != nil {
+		return fmt.Errorf("failed to initialize connection to Cellery Registry %v", err)
+	}
+	imageName := fmt.Sprintf("%s/%s:%s", parsedCellImage.Organization, parsedCellImage.ImageName,
+		parsedCellImage.ImageVersion)
+	repository := parsedCellImage.Organization + "/" + parsedCellImage.ImageName
+	// Creating the Cell Image Digest (Docker file Layer digest)
+	hash := sha256.New()
+	hash.Write(fileBytes)
+	sha256sum := hex.EncodeToString(hash.Sum(nil))
+	cellImageDigest := digest.NewDigestFromEncoded(digest.SHA256, sha256sum)
+
+	// Checking if the the Cell Image already exists in the registry
+	fmt.Println(fmt.Sprintf("\nChecking if the image %s already exists in the Registry", util.Bold(imageName)))
+	cellImageDigestExists, err := hub.HasBlob(repository, cellImageDigest)
+	if err != nil {
+		return err
+	}
+
+	// Pushing the cell image if it is not already uploaded
+	if !cellImageDigestExists {
+		fmt.Println(fmt.Sprintf("Pushing image %s", util.Bold(imageName)))
+		// Read stream of files
+		err = hub.UploadBlob(repository, cellImageDigest, bytes.NewReader(fileBytes), nil)
+		if err != nil {
+			return err
+		}
+	} else {
+		fmt.Println(fmt.Sprintf("\nUsing already existing image %s in %s Registry", util.Bold(imageName),
```

Comment:
Use the stdout provided by Out() function in registry interface to print.

Reply:
Updated.